metaflow/plugins/kfp/kfp.py

```diff
@@ -1,86 +1,293 @@
+    # Level-order traversal is adopted to keep the task-ids in line with what happens during a local metaflow execution.
+    # It is not entirely necessary to keep this order of task-ids if we are able to point to the correct input-paths for
+    # each step. But, using this ordering does keep the organization of data more understandable and natural (i.e., `start`
+    # gets a task id of 1, next step gets a task id of 2 and so on with 'end' step having the highest task id.
+    """
+
+    """
+    Returns the python command template to be used for each step.
+    
+    This method returns a string with placeholders for `datastore_root` and `run_id`
+    which get populated based on the outputs of our initial setup (i.e., when these values are known) in 
+    the pipeline.
+    The rest of the command string is populated using the passed arguments which are known before the run starts.
+    
+    An example constructed command template (to run a step named `hello`):
+    "python downloaded_flow.py --datastore s3 --datastore-root {ds_root} " \
+                     "step hello --run-id {run_id} --task-id 2 " \
+                     "--input-paths {run_id}/start/1"
+    """
+    def build_cmd_template(step_name, task_id, input_paths):
+        python_cmd = "python {downloaded_file_name} --datastore s3 --datastore-root {{ds_root}} " \
+                     "step {step_name} --run-id {{run_id}} --task-id {task_id} " \
+                     "--input-paths {input_paths}".format(downloaded_file_name=DEFAULT_DOWNLOADED_FLOW_FILENAME,
+                                                            step_name=step_name, task_id=task_id, input_paths=input_paths)
+        return python_cmd
+
+    steps_deque = deque(['start']) # deque to process the DAG in level order
+    cur_task_id = 0
+
+    # set of seen steps, i.e., added to the queue for processing
+    seen_steps = set(['start'])
+    # Mapping of steps to task ids
+    task_id_map = {}
+    # Mapping of steps to their command templates
+    command_template_map = {}
+
+    while len(steps_deque) > 0:
+        cur_step = steps_deque.popleft()
+        cur_task_id += 1
+        task_id_map[cur_step] = cur_task_id
+        cur_node = graph.nodes[cur_step]
+
+        # Generate the correct input_path for each step. Note: input path depends on a step's parents (i.e., in_funcs)
+        # Format of the input-paths for reference:
+        # non-join nodes: "run-id/parent-step/parent-task-id",
+        # branch-join node: "run-id/:p1/p1-task-id,p2/p2-task-id,..."
+        # foreach node: TODO: foreach is not considered here
+        if cur_task_id == 1: # start step
+            cur_input_path = '{run_id}/_parameters/0' # this is the standard input path for the `start` step
+        else:
+            if cur_node.type == 'join':
+                cur_input_path = '{run_id}/:'
+                for parent in cur_node.in_funcs:
+                    cur_input_path += parent + '/' + str(task_id_map[parent]) + ','
```

Comment:
please use `"...".format()` string for string construction.

Reply:
Changed